include/gbinder_eventloop.h

```diff
@@ -0,0 +1,189 @@
+     * argument is not NULL. Note that timeout is automatically destroyed if
+     * the callback function returns G_SOURCE_REMOVE.
+     */
+    void (*timeout_remove)(GBinderEventLoopTimeout* timeout);
+
+    /**
+     * callback_new
+     *
+     * Creates a callback object. It returns you a reference, you must
+     * eventually pass the returned object to callback_unref to drop
+     * this reference.
+     *
+     * Note that it doesn't automatically schedule the callback. You
+     * must explicitly call callback_schedule to actually schedule it.
+     * The finalize function is invoked regardless of whether callback
+     * was cancelled or not.
+     */
+    GBinderEventLoopCallback* (*callback_new)(GBinderEventLoopCallbackFunc fun,
+        gpointer data, GDestroyNotify finalize);
+
+    /**
+     * callback_ref
+     *
+     * Increments the reference count. That prevents the object from being
+     * deleted before you drop this reference. The caller makes sure that
+     * argument is not NULL.
+     */
+    void (*callback_ref)(GBinderEventLoopCallback* cb);
+
+    /**
+     * callback_unref
+     *
+     * Decrements the reference count (drops the reference). When reference
+     * count reaches zero, the object gets deleted. The caller makes sure
+     * that argument is not NULL.
+     *
+     * Note that calling callback_schedule temporarily adds an internal
+     * reference until the callback is invoked or callback_cancel is called,
+     * whichever happens first.
+     */
+    void (*callback_unref)(GBinderEventLoopCallback* cb);
+
+    /**
+     * callback_schedule
+     *
+     * Schedules the callback to be invoked in the main loop at some point
+     * in the future (but as soon as possible). The caller makes sure that
+     * argument is not NULL.
+     *
+     * This adds an internal reference to the GBinderEventLoopCallback object
+     * until the callback is invoked or callback_cancel is called, whichever
+     * happens first.
+     */
+    void (*callback_schedule)(GBinderEventLoopCallback* cb);
+
+    /**
+     * callback_cancel
+     *
+     * Makes sure that callback won't be invoked (if it hasn't been
+     * invoked yet) and drops the internal reference. Does nothing
+     * if the callback has already been invoked. The caller makes sure that
+     * argument is not NULL.
+     */
+    void (*callback_cancel)(GBinderEventLoopCallback* cb);
+
+    /**
+     * cleanup
+     *
+     * This function is called when event loop integration is being replaced
+     * with a different one, or libgbinder is being unloaded.
+     */
+    void (*cleanup)(void);
+
+    /* Padding for future expansion */
+    void (*_reserved1)(void);
+    void (*_reserved2)(void);
+    void (*_reserved3)(void);
+    void (*_reserved4)(void);
+    void (*_reserved5)(void);
+    void (*_reserved6)(void);
+    void (*_reserved7)(void);
+    void (*_reserved8)(void);
+    void (*_reserved9)(void);
+    void (*_reserved10)(void);
```

Comment:
Does not really matter etc, but using version/sizeof in front of the struct would remove need for such reserved slots and still allow for 11th extension too if it ever is needed.

Reply:
I would rather add a version field to the end and keep the dummies :) Dummies are easier to use - all you need is to replace it with a real callback and check it for NULL. Versioning requires casts and keeping track of which callback is present in which version, to me that feels like a more complex solution.

In that sense, dummies + version is a perfect combination - it allows simple extension until it we run out of dummies and then (which will hopefully never happens) we can switch to versioning.

Reply:
I have no idea why one might put version / format specifiers / such in the end if the plan is to have forward compatibility.

And it does not need to get complicated if the only purpose is to handle plugins that happen to have been compiled with old version of header and new stuff is always added at the end:

    int call_bar(foo_t *foo)  {
        if(offsetof(foo_t, bar_cb) >= foo->size || !foo->bar_cb)
            return -EWHATEVER;
        return foo->bar_cb();
    }

But never mind, just reserving slots on purpose is already way better than taking padding bytes into active use as an afterthought ;-)

Reply:
For me it starts with: "having dummies at end is technically just fine and dandy." Then I need to resist urge to follow familiar pattern for things like this which among other things makes the dummy slots unnecessary. Which might be overkill in situations like internally used vtables with some room for future extensions...

Reply:
Doing this kind of things for internal APIs makes no sense, and that's one of the reasons why I separate external headers (which live in `include` directory and describe public API which must be kept backward compatible) from the internal ones in `src`. Such a clear separation helps to develop different patterns for internal and external headers.